### PR TITLE
53 final amendments to DB code

### PR DESCRIPTION
**Pull request by AbiHaddon**
added exception handling, doc string, made sure scripts are pep8 compliant using flake8 and added header to script

### CODE REVIEW

File: src/database_3.0.py [hunk 1]
```diff
@@ -6,218 +6,435 @@
...
+
+            # Check if tables exist before creating them - preventing data loss
+            inspector = inspect(self.engine)
+            if not inspector.has_table('samples') or not inspector.has_table(
+                    'panels'):
+
+                # The create_all method creates the tables in the database
+                self.metadata.create_all(self.engine)
+        except SQLAlchemyError as e:
+            print(f"Error occurred when initialising database: {e}")
```
Comment (Fr3dMer) [inline src/database_3.0.py:109]: Something to log?

File: src/database_3.0.py [hunk 2]
```diff
@@ -6,218 +6,435 @@
 
 
 import sqlalchemy as db
-from sqlalchemy import MetaData, Table, Column, String, Date, Float, LargeBinary, ForeignKey, or_, and_, inspect, DateTime, Integer
-from sqlalchemy.sql import select, func
+from sqlalchemy import MetaData, Table, Column, String, Float, LargeBinary, \
```
Comment (Fr3dMer) [inline src/database_3.0.py:9]: is the slash neccesary?

File: src/database_3.0.py [hunk 3]
```diff
@@ -6,218 +6,435 @@
...
-                gene_details=gene_details,
-                JSON=JSON,
-                BED_file_GrCH37=BED_file_GrCH37,
-                BED_file_GrCH38=BED_file_GrCH38
-            ))
             return True
-        else:
-            print("Panel with the same r_code and version already exists")
+        except SQLAlchemyError as e:
+            print(f"Error occurred when inserting patient record: {e}")
```
Comment (Fr3dMer) [inline src/database_3.0.py:135]: this should be logged

File: src/database_3.0.py [hunk 4]
```diff
@@ -6,218 +6,435 @@
...
+                    JSON=JSON,
+                    BED_file_GrCH37=BED_file_GrCH37,
+                    BED_file_GrCH38=BED_file_GrCH38
+                ))
+                return True
+            else:
+                print("Panel with the same r_code and version already exists")
+                return False
+        except SQLAlchemyError as e:
+            print(f"Error occurred when trying to insert panel record "
```
Comment (Fr3dMer) [inline src/database_3.0.py:232]: this should be logged

File: src/database_3.0.py [hunk 5]
```diff
@@ -6,218 +6,435 @@
...
+                    # Select columns in table using select method
+                    db.select(self.panels.columns.JSON)
+                    .where((self.panels.columns.r_code == r_code) & (
+                            self.panels.columns.version == highest_version)))\
+                    .fetchone()
+                return json_query
+            else:
+                return None
+        except SQLAlchemyError as e:
+            print(f"Error occurred when trying to retrieve"
```
Comment (Fr3dMer) [inline src/database_3.0.py:305]: this should be logged

File: src/database_3.0.py [hunk 6]
```diff
@@ -6,218 +6,435 @@
...
+                self.panels.columns.version,
+                self.panels.columns.gene_details,
+                self.panels.columns.BED_file_GrCH37,
+                self.panels.columns.BED_file_GrCH38)
+                .select_from(self.samples.join(self.panels, join_condition))
+                .where(condition)).fetchall()
+
+            return patient_details
+        except SQLAlchemyError as e:
+            print(f"Error occurred when trying to "
```
Comment (Fr3dMer) [inline src/database_3.0.py:349]: log?

File: src/database_3.0.py [hunk 7]
```diff
@@ -6,218 +6,435 @@
...
+            db_manager.samples.columns)).fetchall()
+        output1 = db_manager.connection.execute(db.select(
+            db_manager.panels.columns)).fetchall()
+        print(output)
+        print(output1)
+
+    except Exception as e:
+        print("An error occurred within the execution block:", str(e))
+    finally:
+        # Close the database connection
```
Comment (Fr3dMer) [inline src/database_3.0.py:438]: This is something to remind me, need to close the db in main!

**Review by Fr3dMer — CHANGES_REQUESTED** [review]
Looks good, a few places where logging could be put but other than that I'm happy

File: src/database_3.0.py [hunk 8]
```diff
@@ -6,218 +6,435 @@
```
Comment (AmyGrimwood) [inline src/database_3.0.py:6]: needs a module docstring to comply with pep8

File: src/database_3.0.py (hunk 1 above)
Comment (AmyGrimwood) [reply src/database_3.0.py:109]: Agree

**Review by AmyGrimwood — CHANGES_REQUESTED** [review]
Looks really good, just a few comments

File: src/database_3.0.py (hunk 8 above)
Comment (AbiHaddon) [reply src/database_3.0.py:6]: image
Already present - being minimised in github for some reason

File: src/database_3.0.py (hunk 2 above)
Comment (AbiHaddon) [reply src/database_3.0.py:9]: changed to parentheses

File: src/database_3.0.py (hunk 1 above)
Comment (AbiHaddon) [reply src/database_3.0.py:109]: added logging

File: src/database_3.0.py (hunk 3 above)
Comment (AbiHaddon) [reply src/database_3.0.py:135]: logging added

File: src/database_3.0.py (hunk 4 above)
Comment (AbiHaddon) [reply src/database_3.0.py:232]: added logging

File: src/database_3.0.py (hunk 5 above)
Comment (AbiHaddon) [reply src/database_3.0.py:305]: logging added

File: src/database_3.0.py (hunk 6 above)
Comment (AbiHaddon) [reply src/database_3.0.py:349]: logging added

**Review by Fr3dMer — APPROVED** [review]
Looks good to me :)

**Pull request merged**